Applications/Spire/Include/Spire/Ui/DropDownMenu2.hpp

```diff
@@ -0,0 +1,133 @@
+#ifndef SPIRE_DROP_DOWN_MENU_2_HPP
+#define SPIRE_DROP_DOWN_MENU_2_HPP
+#include <Beam/SignalHandling/ConnectionGroup.hpp>
+#include <QVBoxLayout>
+#include <QWidget>
+#include "Spire/Spire/Spire.hpp"
+#include "Spire/Ui/CustomQtVariants.hpp"
+#include "Spire/Ui/DropDownItem.hpp"
+#include "Spire/Ui/DropShadow.hpp"
+#include "Spire/Ui/ScrollArea.hpp"
+
+namespace Spire {
+
+  //! Represents a Spire-styled drop down menu.
+  class DropDownMenu2 : public QWidget {
+    public:
+  
+      //! Signals that the current item has changed.
+      /*!
+        \param item The current item.
+      */
+      using CurrentSignal = Signal<void (const QVariant& item)>;
+  
+      //! Signals that an item was hovered.
+      /*!
+        \param item The hovered item.
+      */
+      using HoveredSignal = Signal<void (const QVariant& item)>;
+  
+      //! Signals that an item was selected.
+      /*!
+        \param item The selected item.
+      */
+      using SelectedSignal = Signal<void (const QVariant& item)>;
+  
+      //! Constructs a DropDownMenu.
+      /*!
+        \param items The items to display.
+        \param parent The parent widget. Used to determine the position of the
+               DropDownMenu.
+      */
+      explicit DropDownMenu2(std::vector<DropDownItem*> items,
```

Comment:
Include in this PR changes for the ```DropDownMenuItem``` the idea is that was a base class that has several sub-classes, including one common default implementation such as ```LabelDropDownMenuItem``` that just displays text.

Reply:
So the DropDownItem base class is abstract?

Reply:
```DropDownMenuItem``` is an abstract base class, yes. The ```LabelDropDownMenuItem``` is a basic implementation of it that displays the string representation of the ```QVariant``` associated with it.

Reply:
In the case of an abstract base class, are all declared methods marked virtual?

Reply:
They don't have to be no.

Reply:
Based on the [API](https://gist.github.com/spireken/cc683e3ee5953521e07ed269c140d8b1) it would be beneficial to implement these methods in the base DropDownMenuItem class. If none of the methods are pure virtual should the constructor be made protected to effectively make the class abstract to everything but a subclass?

Reply:
If you can implement the ```DropDownMenuItem``` class with that small of an API then go for it, that's great. You don't need a protected constructor.

I suspect you'll need a protected method to let the class know that its highlighted state has been changed. The default implementation doesn't have to do anything ```{}```.

    protected:
      virtual on_highlight_change(bool is_highlighted);

Reply:
So with this small API just implement the LabelDropdownMenuItem as the base DropDownMenuItem?

There shouldn't be a need for now to notify the subclass that the highlight has changed, because the item would only need a repaint which would be done after the highlight is set/reset.

Reply:
Okay go for it, implement the label item.

Reply:
I'm still confused. Just make the DropDownMenuItem the labeled item, or implement a separate LabelDropDownMenuItem? 

If a separate LabelDropDownMenuItem is created and the DropDownMenuItem's constructor isn't protected, it would technically be possible to instantiated a DropDownMenuItem that would not display anything.

Reply:
Implement the ```DropDownMenuItem``` and a subclass ```LabelDropDownMenuItem```. Feel free to make the constructor protected.